langroid/agent/special/sql_chat_agent.py

```diff
@@ -0,0 +1,293 @@
+    column name and its value is a string description of that column.
+    - The 'relationships' key corresponds to another dictionary where each key 
+    is another table name and the value is a description of the relationship to 
+    that table.
+
+    For example:
+    {
+        'table1': {
+            'description': 'description of table1',
+            'columns': {
+                'column1': 'description of column1 in table1',
+                'column2': 'description of column2 in table1'
+            }
+        },
+        'table2': {
+            'description': 'description of table2',
+            'columns': {
+                'column3': 'description of column3 in table2',
+                'column4': 'description of column4 in table2'
+            }
+        }
+    }
+    """
+
+    llm: OpenAIGPTConfig = OpenAIGPTConfig(
+        type="openai",
+        chat_model=OpenAIChatModel.GPT4,
+        completion_model=OpenAIChatModel.GPT4,
+    )
+    prompts: PromptsConfig = PromptsConfig(
+        max_tokens=1000,
+    )
+
+
+class RunQueryTool(ToolMessage):
+    request: str = "run_query"
+    purpose: str = """
+            To run <query> on the database 'db' and 
+            return the results to answer a question.
+            """
+    query: str
+
+
+class SQLChatAgent(ChatAgent):
+    """
+    Agent for chatting with a collection of documents.
+    """
+
+    def __init__(self, config: SQLChatAgentConfig):
+        super().__init__(config)
+        self.config: SQLChatAgentConfig = config
+        self.engine = create_engine(config.database_uri)
+        self.Session = sessionmaker(bind=self.engine)
+        self.metadata = MetaData()
+        self.metadata.reflect(self.engine)
+        self.tables_info = ""
+
+        # User-provided context descriptions
+        context_descriptions = config.context_descriptions
+
+        logger.info(
+            f"""SQLChatAgent initialized with database: 
+            {self.engine} and tables: 
+            {self.metadata.tables}
+            """
+        )
+
+        # Create inspector
+        inspector = inspect(self.engine)
+
+        # Gather table and column information
+        self.tables_info = self.gather_table_info(context_descriptions, inspector)
+
+        # Update the system message with the table information
+        self.config.system_message = self.config.system_message.format(
+            tables=self.tables_info
+        )
+
+        # Enable the agent to use and handle the RunQueryTool
+        self.enable_message(RunQueryTool)
+
+    def agent_response(
+        self,
+        msg: Optional[str | ChatDocument] = None,
+    ) -> Optional[ChatDocument]:
+        # Your override code here
+        if msg is None:
+            return None
+
+        results = self.handle_message(msg)
+        if results is None:
+            return None
+
+        output = results
+        if "There was an error in your SQL Query" in output:
```

Comment:
Should we do a broader match? Maybe the error msg can be different depending on which type of SQL db, etc.

Reply:
This is actually a hard coded string in [response](https://github.com/rithwikbabu/langroid/blob/aa611bb619b2371a700270c48d0a962991e2e2a3/langroid/agent/special/sql_chat_agent.py#L244) to a query error.

Reply:
Ah yes right, then this is ok. l suggest defining a constant like `SQL_ERROR_MSG` since we don't want to have to type the exact same string in two parts of the code (it's fragile, you might mis-type a space, or character, etc). And then use this constant in the two places.

Reply:
Will do, thanks for the suggestion